File: BraintreeCore/src/test/java/com/braintreepayments/api/CrashReporterUnitTest.java
package com.braintreepayments.api;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.lang.ref.WeakReference;

import static junit.framework.Assert.assertSame;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class CrashReporterUnitTest {

    private Thread.UncaughtExceptionHandler defaultExceptionHandler;
    private BraintreeClient braintreeClient;

    @Before
    public void setup() {
        braintreeClient = mock(BraintreeClient.class);
        defaultExceptionHandler = mock(Thread.UncaughtExceptionHandler.class);
    }

    @Test
    public void start_setsSelfAsThreadDefaultExceptionHandler() {
        CrashReporter sut = new CrashReporter(braintreeClient);

        sut.start();
        assertSame(sut, Thread.getDefaultUncaughtExceptionHandler());
    }

    @Test
    public void uncaughtExceptionHandler_whenClientReferenceNull_forwardsInvocationToDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler);

        Thread thread = mock(Thread.class);
        Exception exception = new Exception("error");

        CrashReporter sut = new CrashReporter(new WeakReference<BraintreeClient>(null));
        sut.start();
        sut.uncaughtException(thread, exception);

        verify(defaultExceptionHandler).uncaughtException(thread, exception);
    }

    @Test
    public void uncaughtExceptionHandler_whenClientReferenceNull_restoresOriginalDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler);

        CrashReporter sut = new CrashReporter(new WeakReference<BraintreeClient>(null));
        sut.start();
        sut.uncaughtException(mock(Thread.class), new Exception("error"));

        assertSame(defaultExceptionHandler, Thread.getDefaultUncaughtExceptionHandler());
    }

    @Test
    public void uncaughtExceptionHandler_whenCauseUnknown_forwardsInvocationToDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler);

        Thread thread = mock(Thread.class);
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("test", "test", "test", 1) });

        CrashReporter sut = new CrashReporter(braintreeClient);
        sut.start();
        sut.uncaughtException(thread, exception);

        verify(defaultExceptionHandler).uncaughtException(thread, exception);
    }

    @Test
    public void uncaughtExceptionHandler_whenBraintreeInStackTrace_reportsCrashAndForwardsInvocationToDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler);

        Thread thread = mock(Thread.class);
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("com.braintreepayments.api.CrashReporting", "test", "test", 1) });

        CrashReporter sut = new CrashReporter(braintreeClient);
        sut.start();
        sut.uncaughtException(thread, exception);

        verify(braintreeClient).reportCrash();
        verify(defaultExceptionHandler).uncaughtException(thread, exception);
    }

    @Test
    public void uncaughtExceptionHandler_whenPayPalInStackTrace_reportsCrashAndForwardsInvocationToDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler);

        Thread thread = mock(Thread.class);
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("com.paypal.CrashReporting", "test", "test", 1) });

        CrashReporter sut = new CrashReporter(braintreeClient);
        sut.start();
        sut.uncaughtException(thread, exception);

        verify(braintreeClient).reportCrash();
        verify(defaultExceptionHandler).uncaughtException(thread, exception);
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/CrashReporterUnitTest.kt
package com.braintreepayments.api

import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert.assertSame
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.lang.ref.WeakReference

@RunWith(RobolectricTestRunner::class)
class CrashReporterUnitTest {

    private lateinit var braintreeClient: BraintreeClient
    private lateinit var defaultExceptionHandler: Thread.UncaughtExceptionHandler

    @Before
    fun setup() {
        braintreeClient = mockk(relaxed = true)
        defaultExceptionHandler = mockk(relaxed = true)
    }

    @Test
    fun start_setsSelfAsThreadDefaultExceptionHandler() {
        val sut = CrashReporter(braintreeClient)
        sut.start()
        assertSame(sut, Thread.getDefaultUncaughtExceptionHandler())
    }

    @Test
    fun uncaughtExceptionHandler_whenClientReferenceNull_forwardsToExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception("error")

        val sut = CrashReporter(WeakReference(null))
        sut.start()
        sut.uncaughtException(thread, exception)
        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }

    @Test
    fun uncaughtExceptionHandler_whenClientReferenceNull_restoresOriginalDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val sut = CrashReporter(WeakReference(null))
        sut.start()
        sut.uncaughtException(mockk(), Exception("error"))
        assertSame(defaultExceptionHandler, Thread.getDefaultUncaughtExceptionHandler())
    }

    @Test
    fun uncaughtExceptionHandler_whenCauseUnknown_forwardsInvocationToDefaultExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception()
        exception.stackTrace =
            arrayOf(StackTraceElement("test",
                "test",
                "test",
                1))
        val sut = CrashReporter(braintreeClient)
        sut.start()
        sut.uncaughtException(thread, exception)

        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }

    @Test
    fun uncaughtExceptionHandler_whenBraintreeInStackTrace_reportsCrashToExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception()
        exception.stackTrace = arrayOf(
            StackTraceElement("com.braintreepayments.api.CrashReporting",
                "test",
                "test",
                1)
        )
        val sut = CrashReporter(braintreeClient)
        sut.start()
        sut.uncaughtException(thread, exception)

        verify { braintreeClient.reportCrash() }
        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }

    @Test
    fun uncaughtExceptionHandler_whenPayPalInStackTrace_reportsCrashToExceptionHandler() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception()
        exception.stackTrace = arrayOf(
            StackTraceElement("com.paypal.CrashReporting",
                "test",
                "test",
                1)
        )
        val sut = CrashReporter(braintreeClient)
        sut.start()
        sut.uncaughtException(thread, exception)

        verify { braintreeClient.reportCrash() }
        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }
}
